strawberryfields/tdm/tdmprogram.py

```diff
@@ -291,15 +312,26 @@ def __exit__(self, ex_type, ex_value, ex_tb):
 
         if ex_type is None:
             self.spatial_modes = validate_measurements(self.circuit, self.N)
-            self.construct_circuit()
+            self.timebins = len(self.tdm_params[0])
+            self.total_timebins = self.timebins * self.copies
+            self.rolled_circuit = self.circuit.copy()
+
+    @property
+    def parameters(self):
+        return dict(zip([i.name for i in self.loop_vars], self.tdm_params))
 
-    def construct_circuit(self):
+    def roll(self):
+        """Represent the in a compressed way using without rolling the for loops"""
+        self.circuit = self.rolled_circuit
+        return self
+
+    def unroll(self):
         """Construct program with the register shift"""
-        cmds = self.circuit.copy()
-        self.circuit = []
+        if self.unrolled_circuit is not None:
+            self.circuit = self.unrolled_circuit
+            return self
```

Comment:
> We modify the `TDMProgram`

Is this a possible step? The assumption with the current workflow is that it is not possible to modify the program (at least, in a user-facing manner) after creation. If it turns out it is possible to modify the program after creation, a simply solution is to just `lock()` the program when it exits the `with` context.

Reply:
@antalszava @josh146 It's actually possible to change the `TDMProgram` after creation by appending more operations to the program. They are, however, _not_ represented in the unrolled circuit if the program has already been unrolled earlier, and the new gates will "disappear" if unrolling again (since the unrolling only used `unrolled_circuit` to unroll it if it's already been done before).

The program gets locked when the engine is run, so after that it's no longer possible.

I guess this could be fixed either by forcing an unrolling each time an operation is appended, or simply by locking the circuit early (I think the first option seems nice, and seems fairly simple to implement). Perhaps we could add this as an issue and fix it in a separate PR later?